cdap-hbase-compat-0.98/src/main/java/co/cask/cdap/data2/transaction/messaging/coprocessor/hbase98/MessageTableRegionObserver.java

```diff
@@ -0,0 +1,228 @@
+import java.util.concurrent.TimeUnit;
+
+/**
+ * RegionObserver for the Message Table of Transactional Messaging Service.
+ *
+ * This region observer does row eviction during flush time and compact time by using
+ *
+ * i) Time-To-Live (TTL) information in MetadataTable to determine if a message has expired its TTL.
+ * ii) Generation information in MetadataTable to determine if the message belongs to older generation.
+ * iii) Mark data entries with invalidated transactions (loaded from tx.snapshot) so that invalid list can be pruned.
+ *
+ */
+public class MessageTableRegionObserver extends BaseRegionObserver {
+
+  private static final Log LOG = LogFactory.getLog(MessageTableRegionObserver.class);
+  private static final Gson GSON = new Gson();
+  private static final Type MAP_TYPE = new TypeToken<SortedMap<String, String>>() { }.getType();
+
+  private static final byte[] COL_FAMILY = MessagingUtils.Constants.COLUMN_FAMILY;
+  private static final byte[] COL = MessagingUtils.Constants.METADATA_COLUMN;
+
+  private static int prefixLength;
+  private static LoadingCache<ByteBuffer, Map<String, String>> topicCache;
+  private static TransactionStateCache txStateCache;
+
+  private HTableInterface metadataTable;
+  private CConfigurationReader cConfReader;
+
+  @Override
+  public void start(CoprocessorEnvironment env) throws IOException {
+    if (env instanceof RegionCoprocessorEnvironment) {
+      HTableDescriptor tableDesc = ((RegionCoprocessorEnvironment) env).getRegion().getTableDesc();
+      String metadataTableNamespace = tableDesc.getValue(Constants.MessagingSystem.HBASE_METADATA_TABLE_NAMESPACE);
+      String metadataTableName = tableDesc.getValue(Constants.MessagingSystem.HBASE_METADATA_TABLE_NAME);
+      String hbaseNamespacePrefix = tableDesc.getValue(Constants.Dataset.TABLE_PREFIX);
+      prefixLength = Integer.valueOf(tableDesc.getValue(
+        Constants.MessagingSystem.HBASE_MESSAGING_TABLE_PREFIX_NUM_BYTES));
+      HTableNameConverter nameConverter = new HTable98NameConverter();
+      metadataTable = env.getTable(nameConverter.toTableName(hbaseNamespacePrefix,
+                                                             TableId.from(metadataTableNamespace, metadataTableName)));
+      String sysConfigTablePrefix = nameConverter.getSysConfigTablePrefix(hbaseNamespacePrefix);
+      cConfReader = new CConfigurationReader(env.getConfiguration(), sysConfigTablePrefix);
+      final Long metadataCacheExpiry = Long.parseLong(cConfReader.read().get(
+        Constants.MessagingSystem.COPROCESSOR_METADATA_CACHE_EXPIRATION_SECONDS));
+      topicCache = CacheBuilder.newBuilder()
+        .expireAfterWrite(metadataCacheExpiry, TimeUnit.SECONDS)
+        .maximumSize(1000)
+        .build(new CacheLoader<ByteBuffer, Map<String, String>>() {
+
+          @Override
+          public Map<String, String> load(ByteBuffer topicBytes) throws Exception {
+            Get get = new Get(topicBytes.array());
```

Comment:
This part is dangerous because the part that constructs this `ByteBuffer` is separated from here and if there is any change in the construction, this will break. You can use `array()` safe if and only if `buffer.remaining() == buffer.array().length`